drivers/input/input_mcux_kpp.c

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2025, NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/input/input.h>
+#include <zephyr/logging/log.h>
+#include <fsl_kpp.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/dt-bindings/clock/imx_ccm.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/sys/util.h>
+
+LOG_MODULE_REGISTER(kpp, CONFIG_INPUT_LOG_LEVEL);
+
+#define DT_DRV_COMPAT nxp_mcux_kpp
+
+#define INPUT_KPP_COLUMNNUM_MAX  KPP_KEYPAD_COLUMNNUM_MAX
+#define INPUT_KPP_ROWNUM_MAX     KPP_KEYPAD_ROWNUM_MAX
+
+struct kpp_config {
+	KPP_Type *base;
+	const struct device *ccm_dev;
+	clock_control_subsys_t clk_sub_sys;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct kpp_data {
+	uint32_t clock_rate;
+	struct k_work_delayable work;
+	uint8_t read_keys_old[KPP_KEYPAD_COLUMNNUM_MAX];
+	uint8_t read_keys_new[KPP_KEYPAD_COLUMNNUM_MAX];
+	uint8_t key_pressed_number;
+	const struct device *dev;
+};
+
+static void get_source_clk_rate(const struct device *dev, uint32_t *clk_rate)
+{
+	const struct kpp_config *dev_cfg = dev->config;
+	const struct device *ccm_dev = dev_cfg->ccm_dev;
+	clock_control_subsys_t clk_sub_sys = dev_cfg->clk_sub_sys;
+
+	if (!device_is_ready(ccm_dev)) {
+		LOG_ERR("CCM driver is not installed");
+		*clk_rate = 0;
+		return;
+	}
+
+	clock_control_get_rate(ccm_dev, clk_sub_sys, clk_rate);
+}
+
+static void input_kpp_data_analyze(const struct device *dev)
+{
+	const struct kpp_config *config = dev->config;
+	struct kpp_data *drv_data = dev->data;
+	uint8_t read_keys_new[KPP_KEYPAD_COLUMNNUM_MAX];
+	uint8_t read_keys_old[KPP_KEYPAD_COLUMNNUM_MAX];
+	int row = -1;
+	int col = -1;
+
+	memcpy(read_keys_new, drv_data->read_keys_new, sizeof(read_keys_new));
+	memcpy(read_keys_old, drv_data->read_keys_old, sizeof(read_keys_old));
+
+	for (int i = 0; i < INPUT_KPP_COLUMNNUM_MAX; i++) {
+		if (read_keys_old[i] != read_keys_new[i]) {
+			for (int j = 0; j < INPUT_KPP_ROWNUM_MAX; j++) {
+				if (((read_keys_old[i] ^ read_keys_new[i]) & (1 << j)) != 0) {
+					col = i;
+					row = j;
+				}
+			}
+			drv_data->read_keys_old[i] = drv_data->read_keys_new[i];
+		}
+	}
```

Comment:
don't quite follow the logic here: scan the array, if they are different find if it's a press or a release and send the set of report directly? why look for the last event and only process one?